File: archive/v/visual-basic/bubble-sort.vb
Public Module BubbleSort
    Public Sub Main(args As String())
        Try
            ' Convert string to array of integers
            Dim listOfStringInputs As String() = args(0).Split(",")
            Dim n as Integer = UBound(listOfStringInputs)
            If n < 2
                Usage()
                Exit Sub
            End If
    
            Dim sortArray(n) As Integer
            For i As Integer = 0 To n
                sortArray(i) = Integer.Parse(listOfStringInputs(i))
            Next i

            ' Sort array
            BubbleSort(sortArray)
            
            ' Display array
            Dim first As Boolean = True
            For i As Integer = 0 To n
                If i <> 0
                    System.Console.Write(", ")
                End If
                    
                System.Console.Write(sortArray(i))
            Next i
            System.Console.WriteLine()
        Catch e As FormatException
            Usage()
        Catch e As IndexOutOfRangeException
            Usage()
        End Try
    End Sub

    Public Sub Usage()
        System.Console.WriteLine("Usage: please provide a list of at least two integers to sort in the format ""1, 2, 3, 4, 5""")
    End Sub

    Public Sub BubbleSort(ByRef sortArray As Integer())
        Dim holdInt As Integer
        Dim n As Integer = UBound(sortArray)
        For i As Integer = 0 To n - 1
            For x As Integer = i + 1 To n
                If sortArray(x) < sortArray(i) Then
                    holdInt = sortArray(x)
                    sortArray(x) = sortArray(i)
                    sortArray(i) = holdInt
                End If
            Next x
        Next i
    End Sub
End Module

File: archive/t/typescript/bubble-sort.ts
function DisplayErrorMessage() {
    console.log(
        `Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"`,
    );
    process.exit(1);
}

/**
 * Processes user arguments from CLI, displays usages message and exits if arguments are incorrect
 * @returns user input argument as a list to be merge sorted
 */
function VerifyUserInput(): number[] {
    if (process.argv.length != 3) {
        DisplayErrorMessage();
    }
    // regex string to match requirements
    const regexStr = /^([0-9]+)((,  ?)([0-9]+))+$/;
    const regex = new RegExp(regexStr);
    const user_arg_pos = 2;
    const numbers: string = process.argv[user_arg_pos];

    if (numbers.length == 0 || !numbers.match(regex)) {
        DisplayErrorMessage();
    }
    // split numbers into array and parse values
    return numbers.split(",").map((x) => parseInt(x));
}

/**
 * Swaps the values of the array at the given positions
 * @param array the array with the values to swap
 * @param i the index of the first value to swap
 * @param j the index of the second value to swap
 * @returns 0 to signify that a swap happened
 */
function Swap(array: number[], i: number, j: number): number {
    //list deconstruction to swap values
    [array[i], array[j]] = [array[j], array[i]];
    return 0;
}

/**
 * Sorts the given array using the Bubble Sort Algorithm
 * @param array the array to be sorted
 */
function BubbleSort(array: number[]) {
    var end = array.length;
    while (end > 1) {
        var new_end = 0;
        for (var i = 1; i < end; i++) {
            if (array[i - 1] > array[i]) {
                Swap(array, i - 1, i)
                new_end = i
            }
        }
        end = new_end
    }
}

/** start of program */
const input: number[] = VerifyUserInput();
BubbleSort(input);
console.log(input.join(", "));
